### PR TITLE
proposal: fix sequence merging

### PR DESCRIPTION
**Pull request by davidmdm**
The current implementation of Joy's yaml merge does not account for sequences, whether locked or containing locked values.

The current merge implementation is also not particularly easy to follow.

This PR proposes an alternative merge implementation and provides a suite of tests to outline the difference in output between the current and proposed implementation.

The proposed implementation does have one implementation caveat/simplification, which assumes that the only method to lock a node is via lock tags: `!lock`.

If the proposed implementation is greenlighted, work will be done to port it into the main `internal/yml` package and update existing test suites.

### CODE REVIEW

File: internal/yml/promote/merge_test.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+			Dst:      "[]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3]",
+		},
+		{
+			Name:     "locked seq",
+			Src:      "!lock [1, 2, 3]",
+			Dst:      "[]",
+			Joy:      "!lock [1, 2, 3]",
+			Proposal: "!lock [!!int TODO, !!int TODO, !!int TODO]",
```
Comment (j-martin) [inline internal/yml/promote/merge_test.go:61]: What is going on here? Or would we throw an error?

File: internal/yml/promote/merge_test.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+			Dst:      "[4, 5, 6, 7, 8]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3]",
+		},
+		{
+			Name:     "seq with dst longer and lock",
+			Src:      "[1, 2, 3]",
+			Dst:      "[4, 5, 6, 7, !lock 8]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, 2, 3, !lock 8]",
```
Comment (j-martin) [inline internal/yml/promote/merge_test.go:75]: I do agree we need to fix this. The current behaviour is very misleading.

**Review by j-martin — COMMENTED** [review]
> Pending
I do agree we need to fix this. The current behaviour is very misleading.

File: internal/yml/promote/merge_test.go (hunk 1 above)
Comment (davidmdm) [reply internal/yml/promote/merge_test.go:61]: Here the source is a locked sequence and the destination has no items.

So the result is a sequence with the same length as the source but with all items marked as TODO. The !!int tag is there simply because we set the value to TODO but the node was an int.

Alternatively it could be better to set the result to: !!seq TODO

But currently this version is better than writing our locked values into the destination.

File: internal/yml/promote/merge_test.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		},
+		{
+			Name:     "seq with dst inner lock",
+			Src:      "[1, 2, 3]",
+			Dst:      "[4, !lock 5, 6, 7, !lock 8]",
+			Joy:      "[1, 2, 3]",
+			Proposal: "[1, !lock 5, 3, !lock 8]",
```
[the remark this answers is not in the document]
Comment (j-martin) [reply internal/yml/promote/merge_test.go:82]: @ayahajar raises a good point. I don't think we should replace based on the index since that behaviour would be, at least to me, quite confusing.

We could either do: 

1. Insert the locked value where it makes sense: `[1, !lock 5, 2, 3, !lock8]`
1. Insert the locked value at the end: `[1, 2, 3, !lock 5, !lock8]`

I think option 2 is the least surprising, but yeah, we can do a lot of bike-shedding about this whole subject. 😅

If someone cares about the order and exact content, they should lock the whole array.

`!lock[4, 5, 6, 7]`

**Comment by riclima** [comment]
I've read both implementations and I found the rewrite easier to follow. Coupled with the extra feature of being able to merge sequences I think it's a good change to the current implementation after deciding what to do about merge order.

File: internal/yml/promote/merge_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidmdm) [reply internal/yml/promote/merge_test.go:82]: I think in practice, most use-cases will end up locking the entire value. I tend to be pro index merging, because I imagine that if partial locking is important it would generally be like so:

```
key:
  - !lock env-specific
  - alias
  - alias
```

Where we can understand that the first element is meant to be locked across environments.

Also if we push locks to the end of the sequence, we can't correlate items that were pushed from an original source. Imagine this:

`[!lock a]` gets promoted to an env and becomes: `[!lock TODO]` the user fixes up the value `[!lock b]`.
Now if we promote again how do we know that the two values are correlated? Should it become: `[!lock TODO, !lock b]`?

Basically I think that in the same way that nodes are merged in maps by their keys, nodes should be merged in sequences by their index, otherwise a lot of book keeping needs to be done for a very niche scenario. Also when explaining how sequence merging works, its much easier to say "by index" than to explain when locks get moved to the back or are consolidated with source.

[the remark this answers is not in the document]
Comment (davidmdm) [reply internal/yml/promote/merge_test.go:82]: Also, another point, if ordering does matter, then moving locked values to the back is breaking.

[the remark this answers is not in the document]
Comment (j-martin) [reply internal/yml/promote/merge_test.go:82]: > Also, another point, if ordering does matter, then moving locked values to the back is breaking.

Yes. 🙂 

>> If someone cares about the order and exact content, they should lock the whole array.
>> !lock[4, 5, 6, 7]

At this point, I am wondering if @alexstojda @silphid or @riclima have any opinions about ordering.

[the remark this answers is not in the document]
Comment (riclima) [reply internal/yml/promote/merge_test.go:89]: I slept on it and here's my perspective: to me it makes more sense locking an index than locking a value (AKA @davidmdm suggestion).

> and imply the users have checked the src and dst values ahead of time.

Makes sense in general, but in our use case I think they should "know" what the src and dst values are ahead of time just by looking at the src because the blessed pattern we want to incentivize is to have environments be as similar to each other as possible. Having them vary so much that there are complex scenarios that you need to carefully double check before promoting should be considered an anti-pattern and we should steer users away from it. Ideally promoting a sequence from one environment to another with a lock would mean "give me this exact sequence, but I would like to change the second item of the list."

> Also, another point, if ordering does matter, then moving locked values to the back is breaking.

This becomes important because of this point. There are things that we use in the cluster that ordering does matter. Take for instance this somewhat contrived example derived from a deployment in our cluster:

```yaml
readinessProbe:
  exec:
    command:
       - !lock fee
       - -fi
       - !lock fo
       - -fum
```

List ordering does matter, so if whatever reason we want to change the command `fee` for something else on a different environment and/or (more likely) the value for a flag `-fi`, they're gonna be bumped to the back of the line, requiring a manual reordering on every promote instead of just filling in the `TODO` values.

> We could do:
>
> Insert the locked value where it makes sense: `[1, !lock 5, 2, 3, !lock8]`
> Insert the locked value at the end: `[1, 2, 3, !lock 5, !lock8]`

For this argument I'm not entirely sure my understanding is correct so please correct me if I'm wrong:

If order doesn't matter, then correlation becomes important: say I have `[!lock fee, fi, fo, fum]`. I do a first promote of this sequence on Tuesday from `a` to `b`. Now `b` has: `[fi, fo, fum, TODO]`. I replace the `TODO` value for `prod`, check that all the items I want are there, and move on. Notably, because order doesn't matter, I don't bother to reorder the list, so we gather a slight configuration drift. Then on Wednesday I do some more work on other unrelated parts of `a` and promote it to `b`. Surprise, now I have `[fi, fo fum, prod, TODO]`, because we can't correlate back that we already handled that item. Now every time I promote from  have to keep deleting that `TODO`. Maybe one day I forget and that causes an issue, etc.

I think index locking instead of value locking is simpler, easier to explain to others, and will require less cleaning up during promotes, so I vote for that.

[the remark this answers is not in the document]
Comment (davidmdm) [reply internal/yml/promote/merge_test.go:89]: Just like @riclima, I also slept on it. I think that the alternative to index based merging is not feasible (or at least significantly more difficult) because we want our promotions to be `idempotent`.

If I promote qa to prod, there may be some changes and that all well and good. If I promote qa to prod again without any changes to prod, we should not create more changes since we are already meant to be synced. IE multiple promotes of the same env should not create more than one change. Index based merges guarantees that.

[the remark this answers is not in the document]
Comment (j-martin) [reply internal/yml/promote/merge_test.go:89]: Alright, let's do this then.

**Review by silphid — CHANGES_REQUESTED** [review]
Reading all your comments, I see you all make very good points, but may not draw the proper conclusions from them.  What comes out of all this discussion is that sequence merging is extremely unpredictable. Even if we get the algorithm to be consistent, what users will expect is probably not what will happen, and users would probably use that feature incorrectly and/or get different results/side-effects from what they would expect.

So my personal conclusion is that we simply _should not_ support the merging of individual sequence items, as we have no way to correlate those items together. Items will move around in the different environments and trying to correlate them is impossible if those items often have different values (is it the same item with a different value? just asking the question sound a bit stupid, because we realize that in a sequence the only notion of identity is the value itself, so it just does not add up).

The only thing we should support is what some of you proposed of locking the whole sequence. We should validate and prohibit any locking of sequence elements.

I appears very clear in my mind now, I hope it does for you too, but I'd be super happy to have that discussion live with everyone interested! 😁

**Comment by davidmdm** [comment]
@silphid Index based merging has strong ordering guarantees and works as expected. Not doing this means having to tell developers that locks inside sequences are illegal. I think taking this action is a step towards a more consistent merge/promotion experience.

**Comment by silphid** [comment]
Also, I'd be very curious to see real-world examples where per-item locking within a sequence would make sense (including the scenario of modifying the sequences in different environments and wanting to cross-promote them with specific expectations) and where locking the whole sequence would not get the job done or would result in more work/complexity/confusion for the developers than per-item locking.

**Comment by davidmdm** [comment]
@silphid  I agree that it is incredibly niche and we are unlikely to see item-based locking. That being said, currently, item-based locking, and sequence locking doesn't work.

I think while we are fixing this issue, we might as well create a consistent merge implementation instead of creating cases where if you try to lock an item it will fail.

In javascript land which yaml is a part of (superset of JSON which is javascript object notation) keys and indexes are essentially the same thing. I am biased obviously, but I believe the currently proposed solution is consistent and predictable, and does create edge-cases/limitations to the tooling.

**Comment by davidmdm** [comment]
By as expected, I mean that the way it behaves is consistent, predictable, and idempotent.

It is true, that in a real life scenario, if a user updates an item and tries to promote that value but it is not promoted because that item is locked in the destination that may be unexpected to the user but it also immediately makes sense.

I agree that this topic strays far into the field of theory because we do not have examples of use cases of partially locked sequences.

I would like to advocate for having a sensible merge strategy as the feature, rather than disallowing and validating for specific instances of locks.

Although if it helps move us towards unanimity, banning sequence item locks is an acceptable solution!

**Comment by davidmdm** [comment]
Since index based merging of sequences has not passed the bar, I have come up with another sequence merging strategy that i hope (🤞) will work for everyone, and we can avoid banning item locks. 

Essentially we want the following to be true:
- Locked items in destination are never removed
- Locked items in source are never merged
- source items replace destination items

I think this works and is still idempotent since we never cross merge locked items.

The new strategy looks like so:

```go
func mergeSeq(dst, src *yaml.Node) *yaml.Node {
	var (
		maxLen  = max(len(dst.Content), len(src.Content))
		content = make([]*yaml.Node, 0, maxLen)
	)

	for _, node := range dst.Content {
		if isLocked(node) {
			content = append(content, node)
		}
	}

	for _, node := range src.Content {
		if !isLocked(node) {
			content = append(content, node)
		}
	}

	dst.Content = content
	return dst
}
```

Thoughts, @j-martin @silphid @ayahajar @riclima ?

**Comment by riclima** [comment]
@silphid technically speaking, the current implementation's algorithm is _already_ doing index based sequence locking. For instance, in the following scenario:

src:

```yaml
env:
  - name: !lock fee
     value: fi
  - name: fo
    value: fum
```

dst:

```yaml
env:
  - name: fo
     value: fum
  - name: fee
    value: fi
```

Gives us:

```yaml
env:
  - name: !lock TODO
     value: fi
  - name: fo
    value: fum
```

I.e., the position of the locked `name` value is dependent on its map's sequence index. @davidmdm's proposal is consistent with the current algorithm. In the current algorithm, the exact same behaviour can be achieved as follows if the items in the list are a map:

```yaml
env:
  - name: !lock fee
     value: !lock fi
  - name: fo
    value: fum
```

With @davidmdm, that would extend this existing behaviour to scalars as well and it would provide a shortcut for locking a whole map or subtree like so:

```yaml
env:
  - !lock name: fee
     value: fi
  - name: fo
    value: fum
```

**Comment by silphid** [comment]
@davidmdm what about locking in source, and also in both?  Do you have examples of how the results would look like?  Also, I still insist on coming up with real-world examples/scenarios, because any algorithm not taking into account real scenarios would be moot.

**Comment by davidmdm** [comment]
Ok, just because things are getting confusing, let's reiteterate some of our options using an example

```	
Src:                 "[{a: 1}, {b: 2}]",
Dst:                 "[{b: !lock 2}, {a: 1}, !lock {c: 3}]",
Joy:                 "[{a: 1}, {b: 2}]",
IndexMerge:   "[{a: 1, b: !ock 2}, {b: 2}, !lock {c: 3}]"
FlatMerge:      "[!lock {c: 3}, {a: 1}, {b: 2}]     
```

That's a little complicated but I needed to illustrate the difference with items that contain locks as well as locked items.

Here is what the current three things do:

- Joy: Squashes dst regardless of locked items. Is correct if we ban locking items / items containing locks
- Index Merge: Recursively merges items together by index. Supports items containing locked values.
- FlatMerge: Does not do recursion, but preserves locked items in the destination.

**Comment by davidmdm** [comment]
@silphid 

A more concrete (yet fabricated example):

Suppose we have some piece of yaml in QA:

```yaml
spec:
  containers:
    - name: mycontainer
      image: 1.0.0
```

and I promote it to production, and then say I want to lock the version in production:
```yaml
spec:
  containers:
    - name: mycontainer
      image: !lock 1.0.0
```

If I update my QA version to version `2.0.0` and I promote, you might expect the lock in production to protect the version but it does not. You will get:
```yaml
spec:
  containers:
    - name: mycontainer
      image: 2.0.0
```

Index based merging is true recursive merging, and so allows you to protect nested values. That's the real world use case.

Here it is copied from a test case I just created:
```go
		{
			Name: "real world",
			Src: `spec:
  containers:
    - name: mycontainer
      image: 2.0.0
`,
			Dst: `spec:
  containers:
    - name: mycontainer
      image: !lock 1.0.0
`,
			Joy:      "spec:\n    containers:\n        - name: mycontainer\n          image: 2.0.0",
			Proposal: "spec:\n    containers:\n        - name: mycontainer\n          image: !lock 1.0.0",
		},
```

**Comment by silphid** [comment]
@davidmdm just let me know when you did the clean-up of the duplicated code! :)

**Review by silphid — APPROVED** [review]
